src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp

```diff
@@ -16,112 +15,138 @@
 
-        void Reset() override
-        {
-            Initialize();
-        }
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        _events.ScheduleEvent(EVENT_SPELL_BANSHEEWAIL, 1s);
+        _events.ScheduleEvent(EVENT_SPELL_BANSHEECURSE, 11s);
+        _events.ScheduleEvent(EVENT_SPELL_SILENCE, 13s);
+        _events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+    }
 
-        void JustEngagedWith(Unit* /*who*/) override
-        {
-        }
+    void JustDied(Unit* /*killer*/) override
+    {
+        instance->SetData(TYPE_BARONESS, IN_PROGRESS);
+    }
 
-        void JustDied(Unit* /*killer*/) override
-        {
-            instance->SetData(TYPE_BARONESS, IN_PROGRESS);
-        }
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
 
-        void UpdateAI(uint32 diff) override
-        {
-            if (!UpdateVictim())
-                return;
+        _events.Update(diff);
 
-            //BansheeWail
-            if (BansheeWail_Timer <= diff)
-            {
-                if (rand32() % 100 < 95)
-                    DoCastVictim(SPELL_BANSHEEWAIL);
-                //4 seconds until we should cast this again
-                BansheeWail_Timer = 4000;
-            } else BansheeWail_Timer -= diff;
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
 
-            //BansheeCurse
-            if (BansheeCurse_Timer <= diff)
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
             {
-                if (rand32() % 100 < 75)
+                case EVENT_SPELL_BANSHEEWAIL:
+                    DoCastVictim(SPELL_BANSHEEWAIL);
+                    _events.Repeat(4s);
+                    break;
+                case EVENT_SPELL_BANSHEECURSE:
                     DoCastVictim(SPELL_BANSHEECURSE);
-                //18 seconds until we should cast this again
-                BansheeCurse_Timer = 18000;
-            } else BansheeCurse_Timer -= diff;
-
-            //Silence
-            if (Silence_Timer <= diff)
-            {
-                if (rand32() % 100 < 80)
+                    _events.Repeat(18s);
+                    break;
+                case EVENT_SPELL_SILENCE:
                     DoCastVictim(SPELL_SILENCE);
-                //13 seconds until we should cast this again
-                Silence_Timer = 13000;
-            } else Silence_Timer -= diff;
-
-            DoMeleeAttackIfReady();
+                    _events.ScheduleEvent(EVENT_SPELL_SILENCE, 13s);
+                    break;
+                case EVENT_SPELL_POSSESS:
+                    if (Unit* possessTarget = SelectTarget(SELECT_TARGET_RANDOM, 1, 0, true, false))    // Random target to be possessed
+                    {
+                        if (possessTarget->IsAlive())
+                        {
+                            me->CastStop();
+                            DoCast(possessTarget, SPELL_POSSESS);
```

Comment:
some of these should probably be triggered casts... (hey @Keader check the sniff since you already have it)